### DISCUSSION TITLE
How to upgrade version for all unpublished modules within monorepo

### DISCUSSION THREAD

**Discussion by MarkOSullivan94** [invertase/melos#298] [Q&A]
Hey everyone 👋

I've been reading about the `melos version` option which sounds incredibly helpful to automatically update the version in each of the modules `pubspec.yaml` file and then generate the changelog files for these same modules.

Right now within our monorepo we have not yet adopted [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) and I assume because of this, then the changelog files cannot be generated.

However I was still hoping to use the tool to at least upgrade the version number in all of the `pubspec.yaml` files yet it doesn't seem possible to do that either?

I've tried both `melos version` and `melos version --all` with no success.

**melos version**

```
melos version
   └> /Users/mos/IdeaProjects/Work/verasana

No packages were found that required versioning.
Hint: try running "melos list" with the same filtering options to see a list of packages that were included.
Hint: try running "melos version --all" to include private packages.
```

**melos version --all**

```
melos version
   └> /Users/mos/IdeaProjects/Work/verasana

No packages were found that required versioning.
Hint: try running "melos list" with the same filtering options to see a list of packages that were included.
Hint: try running "melos version --all" to include private packages.
```

### TL;DR

**1. Is it not possible to have a generated changelog after running `melos version` unless you're using conventional commits?**

**2. What would cause Melos to be unable to upgrade the version number within `pubspec.yaml` for the modules in the monorepo?**

Apologies if I've misunderstood the functionality offered!

**Comment by blaugold** [comment]
You can use the `--manual-version` option to manually version packages. It is documented [here](https://melos.invertase.dev/commands/version#--manual-version--v). Does that help you?

**Reply by charlesluzar** [reply]
So the [docs on `--manual-version`](https://melos.invertase.dev/commands/version#--manual-version--v) seemed promising:

> Manually specify a version change for a package. _Can be used multiple times._

I tried this:

```
melos version --all --manual-version=<package_one>:0.0.6+2 --manual-version=<package_two>:0.0.6+2 --manual-version=<package_three>:0.0.6+2 --manual-version=<package_four>:0.0.6+2
```

The `pubspec.yaml` files were all updated as expected, but I ended up with a git tag for each package, eg:

```
% git tag

package_one-v0.0.6+2
package_two-v0.0.6+2
package_three-v0.0.6+2
package_four-v0.0.6+2
```

I'd be looking for a single tag that represented the incremental versioning of all packages in the monorepo, eg:

```
% git tag

v0.0.6+2
```

**Reply by blaugold** [reply]
Ah, I think I get it. In case you're familiar with [Lerna's two modes](https://github.com/lerna/lerna#how-it-works), this sounds like you would like to use the locked mode. Melos currently only implements something comparable to Lerna's independent mode.

**Reply by charlesluzar** [reply]
In a nutshell, that is correct!

**Reply by MarkOSullivan94** [reply]
@blaugold any chance of the locked mode being implemented for Melos in the near future? 🤞 If not, where does it stand in priority terms for Melos?

**Reply by blaugold** [reply]
I think it would be a nice feature to have, but this is the first request for it, so it's not at the top of the list. A PR for this would be welcome, though.
